lib/ret_web/controllers/api/v1/oauth_controller.ex

```diff
@@ -0,0 +1,56 @@
+defmodule RetWeb.Api.V1.OAuthController do
+  use RetWeb, :controller
+
+  plug(RetWeb.Plugs.RateLimit when action in [:show])
+
+  def show(conn, %{"type" => "discord", "state" => state, "code" => code}) do
+    {:ok, claims} = Ret.OAuthToken.decode_and_verify(state)
+    %{"hub_sid" => hub_sid} = claims
+
+    %{"id" => discord_user_id, "email" => email, "verified" => verified} =
+      code |> Ret.DiscordClient.get_access_token() |> Ret.DiscordClient.get_user_info()
+
+    hub = Ret.Hub |> Ret.Repo.get_by(hub_sid: hub_sid) |> Ret.Repo.preload(:hub_bindings)
+    url = hub |> Ret.Hub.url_for()
+
+    if verified do
+      account = email |> Ret.Account.account_for_email()
+
+      Ret.Repo.insert!(
+        %Ret.OAuthProvider{source: :discord, account: account, provider_account_id: discord_user_id},
+        on_conflict: :nothing
+      )
+
+      credentials = %{
+        email: email,
+        token: email |> Ret.Account.account_for_email() |> Ret.Account.credentials_for_account()
+      }
+
+      conn
+      |> put_short_lived_cookie("ret-oauth-flow-account-credentials", credentials |> Poison.encode!())
+      |> put_resp_header("location", url)
+      |> send_resp(307, "")
+    else
+      hub_binding = hub.hub_bindings |> Enum.find(&(&1.type == :discord))
+
+      perms_token =
+        %{
+          join_hub:
+            Ret.DiscordClient.member_of_channel?(discord_user_id, hub_binding.community_id, hub_binding.channel_id),
+          kick_users: false
```

Comment:
On reflection, I think I will leave this as it is, since this only applies to discord users without a verified email, it makes sense to limit their permissions. I need to specify `kick_users` here explicitly since this token is passed on to the SFU which expects tokens to have a `kick_users` claim.